lib/modules.nix

```
@@ -780,9 +780,11 @@ let
   evalOptionValue = loc: opt: defs:
     let
       # Add in the default value for this option, if any.
-      defs' =
-          (optional (opt ? default)
-            { file = head opt.declarations; value = mkOptionDefault opt.default; }) ++ defs;
+      defs' = (optional (opt ? default) {
+        file = head opt.declarations;
+        value = (if opt.additive or false then id else mkOptionDefault) opt.default;
```

I don't think additive is quite the right term. Many types are mergeable, and `additive = false` would not negate that. That'd be unexpected.

I think we could perhaps simplify this to:

```suggestion
        value = mkOverride (opt.defaultPriority or defaultPriority) opt.default;
```

This simplifies the implementation a bit, uses the existing terminology (no conflicts there) and is a bit more powerful (for what it's worth).

Yet another possible syntax would be something like `mkOption { baseDefinition = v; }` (instead of `mkOption { default = v; additive = true; }`), but I think I prefer my prior suggestion.

---

`defaultPriority` is actually deprecated, and it's the default priority of `config` assignments, not `mkOptionDefault`. But I get the idea: Have the options specify a priority for the default.

I think neither the `additive` idea nor setting the priority works out in the end, because it breaks down as soon as you have a more deeply-nested value, e.g.

```nix
lib.mkOption {
  default = {
    foo = 0;
    bar.baz = 10;
    bar.qux = 20;
  };
```

While a user can change `foo` by itself, they can't change `bar.baz` without clearing `bar.qux` too.

What we really need is a default that sets it like
```nix
config = {
  foo = lib.mkOptionDefault 0;
  bar = {
    baz = lib.mkOptionDefault 10;
    qux = lib.mkOptionDefault 20;
  };
}
```

And for this I think an additional `lib.mkOption` attribute is justified, something like `recursiveDefault = true`. And it's not an existing module system concept.

Something very important that needs to be thought about though is allowing options to be backwards compatible, because if options all around just start adding `recursiveDefault = true`, it can break user configs.

---

I don't think the `additive` property works out, because it means that users will have to use `lib.mkForce` to change the defaults. What we really need is an option to push down `lib.mkOptionDefault` to the attribute leaves, e.g.

```nix
lib.mkOption {
  default = {
    foo = 0;
    bar.baz = 10;
    baz.qux = 20;
  };
  recursiveDefault = true;
}
```

turns into effectively

```nix
config = {
  foo = lib.mkOptionDefault 0;
  bar.baz = lib.mkOptionDefault 10;
  baz.qux = lib.mkOptionDefault 20;
}
```

This both overriding defaults without `lib.mkForce` and without clearing the rest.

And the `defaultPriority` idea either suffers from the same `lib.mkForce` problem if the priority is 100 (the default `config` priority) and the clearing problem for lower values (and the problem of default `config` values being ignored entirely for higher values).

---

Whatever solution we end up with though, we need to consider a way for options to stay backwards compatible. Because if they just add `recursiveDefault = true`, it can change behaviour.

---

> I don't think the `additive` property works out, because it means that users will have to use `lib.mkForce` to change the defaults.

Well yes, that is the intention ;)

This flag is the alternative for the pattern of setting the additive default in the implementation. You'd also need mkForce in that case.

This flag should only be used where the user could conceivably want the defaults to remain in place in all but extreme edge-cases.

> What we really need is an option to push down `lib.mkOptionDefault` to the attribute leaves

I think for that purpose it'd be preferable to do this:

```Nix
foo = mkOption {
  additive = true; # Or defaultPriority = ...
  default = {
    foo = lib.mkOptionDefault 0;
    bar.baz = lib.mkOptionDefault 10;
    baz.qux = lib.mkOptionDefault 20;
  };
};
```

This strikes me as a lot simpler than a new boolean flag because with the default being "normal" priority (what's the term here actually?), you have full control of the attrs' priorities and can simply do all your fine-grained priority declarations there. Before, any priorities here would have been clamped to the option default priority unconditionally.